asserts/authority_delegation.go

```diff
@@ -0,0 +1,166 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"fmt"
+)
+
+// AuthorityDelegation holds an authority-delegation assertion, asserting
+// that a specified delegated authority can sign a constrained set
+// of assertion for a given account.
+type AuthorityDelegation struct {
+	assertionBase
+	assertionConstraints []*AssertionConstraints
+}
+
+// AccountID returns the account-id of this authority-delegation.
+func (ad *AuthorityDelegation) AccountID() string {
+	return ad.HeaderString("account-id")
+}
+
+// DelegateID returns the delegated account-id for this authority-delegation.
+func (ad *AuthorityDelegation) DelegateID() string {
+	return ad.HeaderString("delegate-id")
+}
+
+// MatchingConstraints returns all the delegation constraints that match the given assertion.
+func (ad *AuthorityDelegation) MatchingConstraints(a Assertion) []*AssertionConstraints {
+	res := make([]*AssertionConstraints, 0, 1)
+	for _, ac := range ad.assertionConstraints {
+		if ac.Check(a) == nil {
+			res = append(res, ac)
+		}
+	}
+	if len(res) == 0 {
+		return nil
+	}
+	return res
+}
+
+// Implement further consistency checks.
+func (ad *AuthorityDelegation) checkConsistency(db RODatabase, acck *AccountKey) error {
+	if !db.IsTrustedAccount(ad.AuthorityID()) {
+		// XXX if this is relaxed then authority-id must otherwise
+		// match account-id
+		return fmt.Errorf("authority-delegation assertion for %q is not signed by a directly trusted authority: %s", ad.AccountID(), ad.AuthorityID())
+	}
+	_, err := db.Find(AccountType, map[string]string{
+		"account-id": ad.AccountID(),
+	})
+	if IsNotFound(err) {
+		return fmt.Errorf("authority-delegation assertion for %q does not have a matching account assertion", ad.AccountID())
+	}
+	if err != nil {
+		return err
+	}
+	_, err = db.Find(AccountType, map[string]string{
+		"account-id": ad.DelegateID(),
+	})
+	if IsNotFound(err) {
+		return fmt.Errorf("authority-delegation assertion for %q does not have a matching account assertion for delegated %q", ad.AccountID(), ad.DelegateID())
+	}
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// sound
+var _ consistencyChecker = (*AuthorityDelegation)(nil)
+
+// Prerequisites returns references to this authority-delegation's prerequisite assertions.
+func (ad *AuthorityDelegation) Prerequisites() []*Ref {
+	return []*Ref{
+		{Type: AccountType, PrimaryKey: []string{ad.AccountID()}},
+		{Type: AccountType, PrimaryKey: []string{ad.DelegateID()}},
+	}
+}
+
+func assembleAuthorityDelegation(assert assertionBase) (Assertion, error) {
+	// account-id and delegate-id are checked by the general
+	// primary key code
+
+	cs, ok := assert.headers["assertions"]
+	if !ok {
+		return nil, fmt.Errorf("assertions constraints are mandatory")
+	}
+	csmaps, ok := cs.([]interface{})
```

Comment:
This assertion doesn't fully match the err message. Is there a reason not to do:

```suggestion
	csmaps, ok := cs.([]map[string]interface{})
```

instead? It would also eliminate the need for the assertion and err check in the loop AFAICT

Reply:
that is not a valid cast, the compiler complaints

Reply:
I can try again in a follow up, we have a couple more places like this in case

Reply:
Sounds good, thank you